Tests/Common/SimulatedDevice.cs

```diff
@@ -262,73 +263,86 @@ private bool HandleJoinAccept(LoRaPayloadJoinAccept payload)
         // Performs join
         public async Task<bool> JoinAsync(TimeSpan? timeout = null)
         {
-            timeout ??= TimeSpan.FromSeconds(30);
-            using var joinCompleted = new SemaphoreSlim(0);
-            using var joinRequest = WaitableLoRaRequest.CreateWaitableRequest(CreateJoinRequest());
-            var joinRequestPayload = (LoRaPayloadJoinRequest)joinRequest.Payload;
-            var joinSuccessful = false;
+            var retryCount = 0;
+            var effectiveTimeout = timeout ?? TimeSpan.FromSeconds(30);
 
-            void OnMessageReceived(object sender, EventArgs<string> response)
+#if DEBUG
+            if (System.Diagnostics.Debugger.IsAttached)
+            {
+                effectiveTimeout = TimeSpan.FromSeconds(60);
+            }
+#endif
+
+            while (!await TryJoinAsync(effectiveTimeout) && retryCount < MaxJoinRetryCount)
             {
-                try
+                await Task.Delay(TimeSpan.FromSeconds(10));
+                ++retryCount;
+            }
+
+            return retryCount < MaxJoinRetryCount;
+
+            async Task<bool> TryJoinAsync(TimeSpan timeout)
+            {
+                using var joinCompleted = new SemaphoreSlim(0);
```

Comment:
Would be simpler to use a `TaskCompletionSource` here.

Reply:
See ec604f7434a965776c7882ec4c41694b675146a0 for a version using `TaskCompletionSource`.